anndata/_io/specs/registry.py

```diff
@@ -33,16 +87,18 @@ def wrapper(g, k, *args, **kwargs):
 
 class IORegistry(object):
     def __init__(self):
-        self.read: Mapping[tuple[str, IOSpec], Callable] = {}
-        self.read_partial: Mapping[Tuple[str, IOSpec], Callable] = {}
-        self.write: Mapping[Union[Type, Tuple[Type, str]], Callable] = {}
+        self.read: dict[tuple[type, IOSpec, frozenset[str]], Callable] = {}
+        self.read_partial: dict[tuple[type, IOSpec, frozenset[str]], Callable] = {}
+        self.write: dict[
+            tuple[type, type | tuple[type, str], frozenset[str]], Callable
+        ] = {}
 
     def register_write(
         self,
-        dest_type,
-        typ: Union[type, tuple[type, str]],
-        spec,
-        modifiers: frozenset(str) = frozenset(),
+        dest_type: type,
+        typ: type | tuple[type, str],
+        spec: IOSpec | Mapping[str, str],
+        modifiers: Iterable[str] = frozenset(),
```

Comment:
frozenset is actually critical here because it's an unordered hashable collection.

Reply:
https://github.com/theislab/anndata/blob/iospec-errors/anndata/_io/specs/registry.py#L104

Reply:
That’s not a conflict, that’s just two separate types.